metronome/checkpointing/models/src/io/iohk/metronome/checkpointing/models/MerkleTree.scala

```diff
@@ -58,9 +61,124 @@ object MerkleTree {
+          case Seq(a) =>
+            // if the element has no pair we hash it with itself
+            MerkleTree(hashFn(a.hash, a.hash), Some(a))
+        }
+        buildTree(paired)
+      }
+    }
+
+    if (elems.isEmpty)
+      MerkleTree(Hash.empty)
+    else
+      buildTree(elems.toSeq.map(MerkleTree(_))).head
+  }
+
+  val concatThenKec256: (Hash, Hash) => Hash =
+    (a, b) => Hash(Keccak256(a ++ b))
+
+  def verifyProof(
+      proof: Proof,
+      root: Hash,
+      leaf: Hash,
+      hashFn: (Hash, Hash) => Hash = concatThenKec256
+  ): Boolean = {
+    def verify(currentHash: Hash, height: Int, siblings: Seq[Hash]): Hash = {
+      if (siblings.isEmpty)
+        currentHash
+      else {
+        val dir = navigateTree(height, proof.leafIndex)
+        val nextHash =
+          if (dir == 0) hashFn(currentHash, siblings.head)
+          else hashFn(siblings.head, currentHash)
+
+        verify(nextHash, height + 1, siblings.tail)
+      }
+    }
+
+    verify(leaf, 1, proof.siblingPath) == root
+  }
+
+  def generateProof(root: MerkleTree, elem: Hash): Option[Proof] = {
+    val leaves    = findLeaves(root)
+    val leafIndex = leaves.indexOf(elem)
+    val height    = leafCountToHeight(leaves.size)
+
+    if (leafIndex < 0)
+      None
+    else {
+      val siblings = findSiblings(root, height, leafIndex)
+      Some(Proof(leafIndex, leaves.size, siblings))
+    }
+  }
+
+  def findLeaves(tree: MerkleTree): Seq[Hash] = tree match {
+    case MerkleTree(h, None, None) =>
+      Seq(h)
+    case MerkleTree(_, left, right) =>
+      left.map(findLeaves).getOrElse(Nil) ++
+        right.map(findLeaves).getOrElse(Nil)
+  }
+
+  private def findSiblings(
+      tree: MerkleTree,
+      height: Int,
+      leafIndex: Int
+  ): IndexedSeq[Hash] = tree match {
+    case MerkleTree(_, None, None) =>
+      IndexedSeq.empty
+
+    case MerkleTree(_, left, right) =>
+      val dir                 = navigateTree(height, leafIndex)
+      val (traverse, sibling) = if (dir == 0) (left, right) else (right, left)
+      val siblingHash         = sibling.map(_.hash).getOrElse(traverse.get.hash)
+
+      findSiblings(traverse.get, height - 1, leafIndex) :+ siblingHash
+  }
+
+  /** `height = ceil(log2(leafCount))`
+    * Empty tree has height == 0
+    */
+  private def leafCountToHeight(leafCount: Int): Int = {
+    import math._
+    ceil(log(leafCount) / log(2)).toInt
+  }
+
+  //TODO: use an ADT for return instead?
+  /** Indicates the tree traversal direction from a given height towards the leaf:
+    *
+    * 0 - traverse left (take right sibling hash as part of the proof)
+    *
+    * 1 - traverse right (take left hash)
+    */
+  private def navigateTree(height: Int, leafIndex: Int): Int =
+    leafIndex >> (height - 1) & 1
 }
+
+//TODO: seperate into `Node` and `Leaf`. This structure does not enforce integrity (left cannot be empty if
+// right is defined). Also doesn't handle empty trees very well.
+case class MerkleTree(
```

Comment:
To enforce integrity, you could turn it into a `sealed abstract case class MerkleTree private(`. 
In that case the compiler wouldn't generate a `copy` and an `apply` method, and you'd have to produce instances smart constructors calling `new MerkleTree(...) {}`, but you could be sure that the integrity is preserved.

Reply:
Would you prefer an ADT of `FullNode`, `LeftNode` (left child only) and `Leaf`?